src/Sylius/Bundle/CoreBundle/Controller/OrderController.php

```diff
@@ -24,6 +28,76 @@ class OrderController extends BaseOrderController
      *
      * @return Response
      */
+    public function completeAction(Request $request)
+    {
+        $configuration = $this->requestConfigurationFactory->create($this->metadata, $request);
+
+        $this->isGrantedOr403($configuration, ResourceActions::UPDATE);
+        $order = $this->findOr404($configuration);
+
+        $form = $this->resourceFormFactory->create($configuration, $order);
+
+        if (in_array($request->getMethod(), ['POST', 'PUT', 'PATCH'], true) && $form->handleRequest($request)->isValid()) {
+            $order = $form->getData();
+
+            $event = $this->eventDispatcher->dispatchPreEvent(ResourceActions::UPDATE, $configuration, $order);
+
+            if ($event->isStopped() && !$configuration->isHtmlRequest()) {
+                throw new HttpException($event->getErrorCode(), $event->getMessage());
+            }
+            if ($event->isStopped()) {
+                $this->flashHelper->addFlashFromEvent($configuration, $event);
+
+                return $this->redirectHandler->redirectToResource($configuration, $order);
+            }
+
+            try {
+                if ($configuration->hasStateMachine()) {
+                    $this->stateMachine->apply($configuration, $order);
+                }
+
+                $this->manager->flush();
+            } catch (OptimisticLockException $exception) {
+                $this->addFlash('error', 'sylius.checkout.complete_error');
+
+                return $this->redirectHandler->redirectToRoute($configuration, 'sylius_shop_homepage');
+            }
+
+            $this->eventDispatcher->dispatchPostEvent(ResourceActions::UPDATE, $configuration, $order);
+
+            $this->getOrderEmailManager()->sendConfirmationEmail($order);
```

Comment:
I know it is tempting to do it here, but then someone will need to override the controller action if he just want to disable it...

The emails should not be sent before flush, so I'm wondering if using state machine callbacks to send them is a good idea... I know it's convenient but perhaps we should use listeners on `post_action` for that and apply it for all emails, @pjedrzejewski?